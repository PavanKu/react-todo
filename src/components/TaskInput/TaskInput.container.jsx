import React from "react";

import { useTodos } from "../Todo";
import TaskInput from "./TaskInput";

function TaskInputContainer() {
  const { addTodo } = useTodos();
  return <TaskInput handleOnEnter={(task) => addTodo(task)} />;
}

export default TaskInputContainer;
