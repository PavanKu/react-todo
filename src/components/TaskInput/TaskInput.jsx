import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";

import StyledTaskInput from "./TaskInput.styled";

export const DEFAULT_VALUE = "";
export const DEFAULT_PLACEHOLDER = "What needs to be done?";

function TaskInput({
  value = DEFAULT_VALUE,
  placeholder = DEFAULT_PLACEHOLDER,
  handleChange,
  handleOnEnter,
}) {
  const [inputValue, setInputValue] = useState(value);

  const handleInputChange = (event) => {
    const text = event.target.value;
    setInputValue(text);
    if (handleChange) {
      handleChange(text);
    }
  };

  const handleFormSubmit = (event) => {
    event.preventDefault();
    if (handleOnEnter) {
      handleOnEnter(inputValue);
      setInputValue(DEFAULT_VALUE);
    }
  };

  useEffect(() => {
    setInputValue(value);
  }, [value]);

  return (
    <form onSubmit={handleFormSubmit}>
      <StyledTaskInput
        name="task-input"
        value={inputValue}
        onChange={handleInputChange}
        placeholder={placeholder}
      />
    </form>
  );
}

TaskInput.propTypes = {
  value: PropTypes.string,
  placeholder: PropTypes.string,
  handleChange: PropTypes.func,
  handleOnEnter: PropTypes.func,
};

export default TaskInput;
