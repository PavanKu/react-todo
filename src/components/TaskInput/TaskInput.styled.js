import styled from "styled-components";

const StyledTaskInput = styled.input.attrs(() => ({
  type: "text",
}))`
  padding: 16px 16px 16px 60px;
  border: none;
  box-shadow: inset 0 -2px 1px rgba(0, 0, 0, 0.03);
  font-size: 24px;
  line-height: 1.4em;
  width: 100%;
  outline: none;
  background: #fefefe;
  box-sizing: border-box;
`;

export default StyledTaskInput;
