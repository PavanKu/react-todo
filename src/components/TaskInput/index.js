import TaskInputContainer from "./TaskInput.container";

export { TaskInputContainer };
