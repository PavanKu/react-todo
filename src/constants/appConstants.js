const TAB_OPTIONS = {
  ALL: "All",
  ACTIVE: "Active",
  COMPLETED: "Completed",
};

export { TAB_OPTIONS };
