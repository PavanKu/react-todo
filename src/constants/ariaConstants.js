const ARIA_ROLE_HEADING = "heading";
const ARIA_ROLE_TEXTBOX = "textbox";
const ARIA_ROLE_CHECKBOX = "checkbox";
const ARIA_ROLE_BUTTON = "button";
const ARIA_ROLE_TABLIST = "tablist";
const ARIA_ROLE_TAB = "tab";

const ARIA_TAB_SELECTED_ATTRIBUTE = "aria-selected";

export {
  ARIA_ROLE_HEADING,
  ARIA_ROLE_TEXTBOX,
  ARIA_ROLE_CHECKBOX,
  ARIA_ROLE_BUTTON,
  ARIA_ROLE_TABLIST,
  ARIA_ROLE_TAB,
  ARIA_TAB_SELECTED_ATTRIBUTE,
};
