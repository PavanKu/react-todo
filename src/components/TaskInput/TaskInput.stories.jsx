import React from "react";
import { storiesOf } from "@storybook/react";
import { withKnobs, text } from "@storybook/addon-knobs";
import { action } from "@storybook/addon-actions";
import TaskInput from "./TaskInput";
import { DEFAULT_PLACEHOLDER } from "./TaskInput";

const SIMPLE_TASK = "Buy Milk";
const AddTaskWithDefaultValue = () => <TaskInput />;
const AddTaskWithValue = () => <TaskInput value={SIMPLE_TASK} />;
const Playground = () => (
  <TaskInput
    placeholder={text("placeholder", DEFAULT_PLACEHOLDER)}
    value={text("value", SIMPLE_TASK)}
    handleChange={action("TaskInput: handleChange:")}
  />
);

storiesOf("Task Input", module)
  .addDecorator(withKnobs)
  .add("Task Input without value", AddTaskWithDefaultValue)
  .add("Task Input with value", AddTaskWithValue)
  .add("Playground", Playground);
