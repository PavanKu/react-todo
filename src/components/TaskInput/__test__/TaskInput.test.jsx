import React from "react";
import { cleanup, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import TaskInput, { DEFAULT_PLACEHOLDER, DEFAULT_VALUE } from "../TaskInput";

import { ARIA_ROLE_TEXTBOX } from "../../../constants/ariaConstants";

describe("<TaskInput />", () => {
  afterEach(cleanup);

  it("should render wihtout crashing", () => {
    render(<TaskInput />);
  });

  it("should render with default values", () => {
    const { getByRole, getByPlaceholderText } = render(<TaskInput />);
    const elementByRole = getByRole(ARIA_ROLE_TEXTBOX);
    const elementByPlaceholder = getByPlaceholderText(DEFAULT_PLACEHOLDER);

    expect(elementByRole).toHaveValue(DEFAULT_VALUE);
    expect(elementByPlaceholder).toEqual(elementByRole);
  });

  it("should render as per props", () => {
    const inputValue = "Buy Milk";
    const placeholder = "What do to like to do today?";

    const { getByRole, getByPlaceholderText } = render(
      <TaskInput value={inputValue} placeholder={placeholder} />
    );

    const elementByRole = getByRole(ARIA_ROLE_TEXTBOX);
    const elementByPlaceholder = getByPlaceholderText(placeholder);

    expect(elementByRole).toHaveValue(inputValue);
    expect(elementByRole).toEqual(elementByPlaceholder);
  });

  it("should change value when user types", () => {
    const inputValue = "Buy Bread";
    const { getByRole } = render(<TaskInput handleChange={() => {}} />);

    const elementByRole = getByRole(ARIA_ROLE_TEXTBOX);

    expect(elementByRole).toHaveValue(DEFAULT_VALUE);
    userEvent.type(elementByRole, inputValue);
    expect(elementByRole).toHaveValue(inputValue);
  });

  it("should call change handler with user input", () => {
    const inputValue = "Buy Toy";
    let userInput;

    const { getByRole } = render(
      <TaskInput handleChange={(value) => (userInput = value)} />
    );
    const elementByRole = getByRole(ARIA_ROLE_TEXTBOX);

    expect(userInput).toBeUndefined();
    userEvent.type(elementByRole, inputValue);
    expect(userInput).toStrictEqual(inputValue);
  });

  it("should match snapshot", () => {
    const { container } = render(
      <TaskInput value={DEFAULT_VALUE} placeholder={DEFAULT_PLACEHOLDER} />
    );

    expect(container).toMatchSnapshot();
  });
});
