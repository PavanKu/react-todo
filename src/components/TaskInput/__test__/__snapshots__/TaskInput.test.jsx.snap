// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TaskInput /> should match snapshot 1`] = `
<div>
  <form>
    <input
      class="sc-AxjAm hSKMid"
      name="task-input"
      placeholder="What needs to be done?"
      type="text"
      value=""
    />
  </form>
</div>
`;
